### PR TITLE
FIX: improve symlog ticker

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

`SymLogLocator` is supposed to not show any ticks in the range [-linthres, linthresh], except for zero.  However, the `floor` call can leave a tick inside the lin threshold region, and this usually looks quite bad.  

```python
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots(figsize=(3, 3))
ax.plot(np.linspace(-1, 1, 1000))
ax.set_yscale('symlog', linthresh=0.03)
plt.show()
```

### Before

Before

### After:

After

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by jklymak** [comment]
This is motivated by https://matplotlib.org/devdocs/gallery/userdemo/colormap_normalizations_symlognorm.html where the colorbar is terrible because of this bug....

**Comment by dstansby** [comment]
I think the better fix is to change `floor` to `ceil` on line 2519 above? (which is what I did at the end of https://github.com/matplotlib/matplotlib/pull/14309/files, which never got merged)

**Comment by jklymak** [comment]
Maybe?   We often return out of bounds ticks so I wasn't sure if that was desired here.

**Comment by dstansby** [comment]
What do you mean by out of bounds ticks? I think changing `floor` to `ceil` will have exactly the same effect as the current diff, and is makes more sense (in my opinion) when reading the code.

**Comment by jklymak** [comment]
Consider 

```python 
import matplotlib.pyplot as plt 
import matplotlib.ticker as mticker

loc = mticker.LogLocator()
print(loc.tick_values(1e-3, 1e3))
```

This returns 

```
[1.e-04 1.e-03 1.e-02 1.e-01 1.e+00 1.e+01 1.e+02 1.e+03 1.e+04]
```

and note that the first and last ticks are (well) outside the vmin and vmax.

**Comment by dstansby** [comment]
Ah I think that's because the `hi` lims should be using `np.floor`, not `np.ceil`? Essentially it looks like both calls should be switched in lines 2519-2520

**Comment by jklymak** [comment]
No, I mean the hi lim uses `ceil` because we typically want one tick _above_ `vmax`.

**Comment by dstansby** [comment]
Sorry, I got a bit confused. to go back to my original suggestion, changing https://github.com/matplotlib/matplotlib/blob/a6e01b5d3c989ba456be2531f608200755081af9/lib/matplotlib/ticker.py#L2519

from `floor` to `ceil` should fix the original issue. Using `floor` currently gives the first tick outside the log range; changing to `ceil` will give the first tick inside the log range.

**Comment by jklymak** [comment]
@dstansby We want one tick below vmin and one tick above vmax, and then we clip them when we draw the axis.

**Comment by dstansby** [comment]
In general yes, but that's not what this PR is about - the vmin/vmax of the Axis are distinct from the linear threshold of the ticker. As you say right at the top:

> SymLogLocator is supposed to not show any ticks in the range [-linthres, linthresh], except for zero. However, the floor call can leave a tick inside the lin threshold region, and this usually looks quite bad.

Changing the https://github.com/matplotlib/matplotlib/blob/a6e01b5d3c989ba456be2531f608200755081af9/lib/matplotlib/ticker.py#L2519

to ceil solves the issue, and is a much simpler and easy to follow piece of code than what is currently proposed.

**Comment by jklymak** [comment]
That same code is used above _and_ below the threshold...

**Comment by dstansby** [comment]
I don't think we're doing a very good job of convincing each other 😄  - I won't block this, but leave for others to review.

**Comment by jklymak** [comment]
Well let me try again: `get_log_range` is used for both range_a and range_c  So the hi part of range_a is thresh, but the hi part of range_c is vmax.  So you would want to use `floor` and for range_a you'd want to use `ceil` for range_c.  We could just get rid of `get_log_range` and write it all out, but that seemed more invasive than clipping the existing result at thresh.

**Comment by dstansby** [comment]
For the a and c ranges the calls are:
```python
get_log_range(abs(a_upper_lim), abs(vmin) + 1)
get_log_range(c_lower_lim, vmax + 1)
```
The first argument for *both* the a and c ranges is `linthresh` (if `linthresh` is within the Axis limits), so applying `ceil` to the first argument will have the correct effect for both the a and c ranges, ie. returning the first tick outside the linear range.

**Comment by jklymak** [comment]
Oh duh, my apologies, you are absolutely correct.  I've fixed it.  Sorry for being so dense...

**Comment by dstansby** [comment]
Don't worry! Sorry for not explaining very well!

**Comment by jklymak** [comment]
Hmm, except that fails `symlog2` test.   Which I agree I've updated, but I think the update is better than before...

### Master
Oldway

### Suggested by @dstansby

dstansby

### My suggestion

Myway

I guess not of these are quite right (i.e. the first axes should have more than just a zero tick!)

**Comment by dstansby** [comment]
Ahh I think this is because of the floating point issues using `np.log` instead of `np.log10` when linthresh lies exactly on a decade. So I think changing `log` to `log10` will fix it for base 10 (in https://github.com/matplotlib/matplotlib/pull/21177 ut seemed to work for other bases too) , but given this maybe it's better to go back to `floor` and clipping the ticks after all...

> I guess not of these are quite right (i.e. the first axes should have more than just a zero tick!)

From memory, in this case the locator should just end up using the default linear locator?

**Comment by tacaswell** [comment]
This PR is affected by a re-writing of our history to remove a large number of accidentally committed files [see discourse](https://discourse.matplotlib.org/t/minor-re-writing-of-history-and-moving-from-master-to-main-for-default-branch/22354/6) for details.

To recover this PR it will need be rebased onto the new default branch (main).  There are several ways to accomplish this, but we recommend (assuming that you call the matplotlib/matplotlib remote `"upstream"`

```bash
git remote update
git checkout main
git merge --ff-only upstream/main
git checkout YOUR_BRANCH
git rebase --onto=main upstream/old_master
# git rebase -i main # if you prefer
git push --force-with-lease   # assuming you are tracking your branch
```

If you do not feel comfortable doing this or need any help please reach out to any of the Matplotlib developers.  We can either help you with the process or do it for you.

Thank you for your contributions to Matplotlib and sorry for the inconvenience.
